docker/docker_client.go

```diff
@@ -212,6 +208,13 @@ func newDockerClientWithDetails(sys *types.SystemContext, registry, username, pa
 		tr.TLSClientConfig.InsecureSkipVerify = true
 	}
 
+	if username == "" && password == "" {
+		username, password, err = config.GetAuthentication(sys, registry)
```

Comment:
The handling of the various `*.docker.io` host names does not look right, or at least it is changing:

- Before, for `newDockerClientFromRef`, `GetAuthentication` was called with `docker.io`; now it is done with `dockerRegistry` == `registry-1.docker.io`.  That could be solved by moving this code a bit higher.
- `SearchRegistry` sets `registry` to `dockerV1Hostname`, so that’s what `GetAuthentication` will be called with. That’s more difficult.

`GetAuthentication` does eventually normalize all three to a single value — but only for the `AuthConfigs` lookup, not for `CredHelpers`; so I think we don’t want to change the inputs.

So, either `newDockerClientWithDetails` could get a separate “semantic host name” / “configuration lookup key” (I don’t know how to call that) parameter, which remains `docker.io`, or we could add a ~duplicate `GetAuthentication` call into `SearchRegistry` instead of consolidating both into `newDockerClientWithDetails`. 

Both are somewhat ugly, the latter one maybe a bit less: the distinction between the use of the three `docker.io` hostnames concisely seems difficult; whereas a simple duplication of a call to an external method is just three extra _simple_ lines and the distinction between the use of the hostnames is captured directly in straight-line code.

Or there well could be some much cleaner way to express it, one I’m missing.

Reply:
This still calls `GetAuthentication` after the two remappings, i.e. with `dockerRegistry` / `dockerV1Hostname`.  (It now won’t affect `newDockerClientFromRef` because that does its own lookup, but `SearchRegistry` will cause `dockerV1Hostname` to be looked up.)

Reply:
(I don’t know, maybe it would make sense for `GetAuthentication` to always normalize these three inputs to the same key; it’s not at all obvious that it should accept all the pre-normalized kinds of input — https://docs.docker.com/engine/reference/commandline/login/#extended-description says that the keys are “registry domains”, which seems to rule out at least the _schema_`://`_host_ format accepted by `normalizeRegistry`.

If `GetAuthentication` mapped all of the different `docker.io`-related hostnames and specs to a single value, the callers in this file could be more lenient and we could have a single `GetAuthentication` call like this one. But I haven’t actually looked into the details of what docker does, nor what the documentation says, for the corner cases.)

Reply:
> says that the keys are “registry domains”

I meant the keys in `credHelpers`.

Reply:
Should `GetAuthentication` be called right at the beginning of the function then? Before any of the remapping is done?

Reply:
That helps with `dockerRegistry`, but not with `dockerV1Hostname`.

Reply:
finally fixed now. Added a secong `GetAuthentication` call in `SearchRegistry` before the registry hostname change happens.